sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala

```diff
@@ -114,4 +114,144 @@ object EstimationUtils {
+   * @param higherEnd a given upper bound value of a specified column value range
+   * @param lowerEnd a given lower bound value of a specified column value range
+   * @param histogram a numeric equi-height histogram
+   * @return the number of bins for column values in [lowerEnd, higherEnd].
+   */
+  def getOccupationBins(
+      higherId: Int,
+      lowerId: Int,
+      higherEnd: Double,
+      lowerEnd: Double,
+      histogram: Histogram): Double = {
+    if (lowerId == higherId) {
+      val curBin = histogram.bins(lowerId)
+      getOccupation(higherEnd, lowerEnd, curBin)
+    } else {
+      // compute how much lowerEnd/higherEnd occupies its bin
+      val lowerCurBin = histogram.bins(lowerId)
+      val lowerPart = getOccupation(lowerCurBin.hi, lowerEnd, lowerCurBin)
+
+      val higherCurBin = histogram.bins(higherId)
+      val higherPart = getOccupation(higherEnd, higherCurBin.lo, higherCurBin)
+
+      // the total length is lowerPart + higherPart + bins between them
+      lowerPart + higherPart + higherId - lowerId - 1
+    }
+  }
+
+  /**
+   * Returns the number of distinct values, ndv, for column values in [lowerEnd, higherEnd].
+   * The column value distribution is saved in an equi-height histogram.
+   *
+   * @param higherId id of the high end bin holding the high end value of a column range
+   * @param lowerId id of the low end bin holding the low end value of a column range
+   * @param higherEnd a given upper bound value of a specified column value range
+   * @param lowerEnd a given lower bound value of a specified column value range
+   * @param histogram a numeric equi-height histogram
+   * @return the number of distinct values, ndv, for column values in [lowerEnd, higherEnd].
+   */
+  def getOccupationNdv(
+      higherId: Int,
+      lowerId: Int,
+      higherEnd: Double,
+      lowerEnd: Double,
+      histogram: Histogram)
+    : Long = {
+    val ndv: Double = if (higherEnd == lowerEnd) {
+      1
+    } else if (lowerId == higherId) {
+      val curBin = histogram.bins(lowerId)
+      getOccupation(higherEnd, lowerEnd, curBin) * curBin.ndv
+    } else {
+      // compute how much the [lowerEnd, higherEnd] range occupies the bins in a histogram.
+      // Our computation has 3 parts: the smallest/min bin, the middle bins, the largest/max bin.
+      val minCurBin = histogram.bins(lowerId)
+      val minPartNdv = getOccupation(minCurBin.hi, lowerEnd, minCurBin) * minCurBin.ndv
+
+      val maxCurBin = histogram.bins(higherId)
+      val maxPartNdv = getOccupation(higherEnd, maxCurBin.lo, maxCurBin) * maxCurBin.ndv
+
+      // The total ndv is minPartNdv + maxPartNdv + Ndvs between them.
+      // In order to avoid counting same distinct value twice, we check if the upperBound value
+      // of next bin is equal to the hi value of the previous bin.  We bump up
+      // ndv value only if the hi values of two consecutive bins are different.
```

Comment:
this doesn't match the code, the actual logic is: `only if the lo and hi values of the bin are different`

Reply:
I will change the comment so that it matches with the code.  Actually my original comment means the same thing as your comment.  This is because the hi value of a bin is equal to the lo value of the next bin.